lib/registry/src/lib.rs

```diff
@@ -1180,6 +1189,176 @@ pub fn get_all_available_registries() -> Result<Vec<String>, String> {
     Ok(registries)
 }
 
+#[derive(Debug, PartialEq, Clone)]
+pub struct RemoteWebcInfo {
+    pub checksum: String,
+    pub manifest: webc::Manifest,
+}
+
+pub fn install_webc_package(url: &Url, checksum: &str) -> Result<(), String> {
+    tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap()
+        .block_on(async { install_webc_package_inner(url, checksum).await })
+}
+
+async fn install_webc_package_inner(url: &Url, checksum: &str) -> Result<(), String> {
+    use futures_util::StreamExt;
+
+    let path = get_webc_dir().ok_or_else(|| "no webc dir".to_string())?;
+
+    let _ = std::fs::create_dir_all(&path);
+
+    let webc_path = path.join(checksum);
+
+    let mut file =
+        std::fs::File::create(&webc_path).map_err(|e| format!("{}: {e}", webc_path.display()))?;
+
+    let client = {
+        let builder = reqwest::Client::builder();
+
+        #[cfg(not(target_os = "wasi"))]
+        let builder = if let Some(proxy) =
+            crate::graphql::proxy::maybe_set_up_proxy().map_err(|e| format!("{e}"))?
+        {
+            builder.proxy(proxy)
+        } else {
+            builder
+        };
+
+        builder.build().map_err(|e| format!("{e}"))?
+    };
+
+    let res = client
+        .get(url.clone())
+        .header(ACCEPT, "application/webc")
+        .send()
+        .await
+        .map_err(|e| format!("{e}"))?;
+
+    let mut stream = res.bytes_stream();
+
+    while let Some(item) = stream.next().await {
+        let item = item.map_err(|e| format!("{e}"))?;
+        file.write_all(&item).map_err(|e| format!("{e}"))?;
+    }
+
+    Ok(())
+}
+
+/// Returns a list of all installed webc packages
+pub fn get_all_installed_webc_packages() -> Vec<RemoteWebcInfo> {
+    let dir = match get_webc_dir() {
+        Some(s) => s,
+        None => return Vec::new(),
+    };
+
+    let read_dir = match std::fs::read_dir(dir) {
+        Ok(s) => s,
+        Err(_) => return Vec::new(),
+    };
+
+    read_dir
+        .filter_map(|r| Some(r.ok()?.path()))
+        .filter_map(|path| {
+            webc::WebCMmap::parse(
+                path,
+                &webc::ParseOptions {
+                    parse_atoms: false,
+                    parse_volumes: false,
+                    ..Default::default()
+                },
+            )
+            .ok()
+        })
+        .filter_map(|webc| {
+            let mut checksum = webc.checksum.as_ref().map(|s| &s.data)?.to_vec();
+            while checksum.last().copied() == Some(0) {
+                checksum.pop();
+            }
```

Comment:
What is this code doing?

Reply:
The `checksum` has a bunch of zeros at the end (it's currently encoded that way in the webc format). This code strips the zeros because otherwise the filename would become too long. So:

`3ea47cb0000000000000` -> `3ea47cb`

Reply:
Split it into a different function